mobile_app_connector/models/account_invoice.py

```diff
@@ -119,50 +62,55 @@ def mobile_post_invoice(self, json_data, **parameters):
 
         # create invoice and merge lines
-        if not invoice:
-            lines_cmd = [(0, 0, v) for v in invoice_lines_values]
-            if sponsorships_lines:
-                lines_cmd = [(6, 0, sponsorships_lines.ids)] + lines_cmd
-                # order is important
-
-            invoice = self.sudo().create({
-                'partner_id': wrapper.partner_id,
-                'invoice_line_ids': lines_cmd,
-                'origin': wrapper.source,
-                'type': 'out_invoice',
-                'date_invoice': fields.Date.today()
-            })
-        elif sponsorships_lines or invoice_lines_values:
-            # add or create the lines in the existing invoice
-
-            if sponsorships_lines:
-                invoice.invoice_line_ids += sponsorships_lines
-            if invoice_lines_values:
-                invoice.write({
-                    'invoice_line_ids': [(0, 0, v) for v in invoice_lines_values]
-                })
-
-        if date_due:
-            invoice.date_due = date_due
-        invoice.origin = wrapper.source
-
-        if notify:
-            partner_id = self.env['staff.notification.settings'].get_param(
-                'gift_notify_ids')[0]
-
-            invoice.message_needaction = True
-            invoice.user_id = \
-                self.env['res.partner'].sudo().browse(partner_id).user_ids[:1]
-            invoice.message_post(
-                _("This invoice created from the app need to be manually processed. "
-                  "It might cancel older invoices."))
-
-        for line in invoice.invoice_line_ids:
-            bckp_price = line.price_unit
-            line._onchange_product_id()
-            line.price_unit = bckp_price
-
-        invoice.action_invoice_open()
+        lines_cmd = [(0, 0, v) for v in invoice_lines_values]
+        existing_invoice = False
+        invoice = self.sudo().create({
+            'partner_id': wrapper.partner_id,
+            'invoice_line_ids': lines_cmd,
+            'origin': wrapper.source,
+            'type': 'out_invoice',
+            'date_invoice': fields.Date.today()
+        })
+
+        if wrapper.sponsorships_payments:
+            # User is paying for sponsorship: only if everything match an open invoice,
+            # we will use it.
+            sponsorship_invoice = invoice.search([
+                ('partner_id', '=', wrapper.partner_id),
+                ('state', '=', 'open'),
+                ('amount_total', '=', invoice.amount_total),
+            ], order='date asc', limit=1)
+            sp_invoice_contracts = sponsorship_invoice.invoice_line_ids.mapped(
+                'contract_id')
+            new_invoice_contracts = invoice.invoice_line_ids.mapped('contract_id')
+            common_sponsorships = sp_invoice_contracts & new_invoice_contracts
+            if common_sponsorships and len(common_sponsorships) == \
+                    len(sp_invoice_contracts):
```

Comment:
You will never match the 50.- (42+8) sponsorships here. They have 2 lines in the original invoice (with the same contract_id) but only one in the new invoice. (the & operator computes a set)
But maybe you did it on purpose...

Reply:
Yes it will match, the set contains unique elements and I'm comparing the related sponsorships. So even if one invoice have multiple lines, if they all are related to the same sponsorship, the resulting set will contain only one element.

Reply:
my bad, I thought `.mapped('contract_id')` was a 1 to 1 mapping